File: gaia-sdk-javascript/src/graphql/GaiaStreamClientBuilder.ts
import {ClientOptions} from '..';
import {GaiaCredentials} from '../api/GaiaCredentials';
import {GaiaStreamClient} from './GaiaStreamClient';
import {StreamHttpTransporter} from '../http/StreamHttpTransporter';
import {IStreamTransporter} from '../api/IStreamTransporter';

export class GaiaStreamClientFactory {
    http(url: string) {
        return GaiaStreamClientBuilder.http(url);
    }
}

export class GaiaStreamClientBuilder {
    transporter: IStreamTransporter;
    credentials?: GaiaCredentials;
    contentType: string = 'application/json';

    constructor(transporter: IStreamTransporter) {
        this.transporter = transporter;
    }

    public static http(url: string) {
        return new GaiaStreamClientBuilder(new StreamHttpTransporter(url));
    }

    public withCredentials(credentials: GaiaCredentials) {
        this.credentials = credentials;
        return this;
    }

    public withContentType(contentType: string) {
        this.contentType = contentType;
        return this;
    }

    public build() {
        const options = new ClientOptions(this.credentials!, this.contentType);
        return new GaiaStreamClient(options, this.transporter);
    }

}

File: gaia-sdk-python/gaia_sdk/http/GaiaStreamClientBuilder.py
from gaia_sdk.api.GaiaCredentials import GaiaCredentials
from gaia_sdk.api.client_options import ClientOptions
from gaia_sdk.http.GaiaStreamClient import GaiaStreamClient
from gaia_sdk.http.HttpTransporter import HttpTransporter

class GaiaStreamClientFactory(object):
    def http(self, url):
        return GaiaHttpStreamClientBuilder(url)


class GaiaHttpStreamClientBuilder(object):

    def __init__(self, url):
        self.url = url

    def with_credentials(self, credentials: GaiaCredentials):
        self.credentials = credentials
        return self

    def build(self):
        return GaiaStreamClient(HttpTransporter(self.url), ClientOptions(self.credentials))
